staging/src/k8s.io/dynamic-resource-allocation/structured/allocator_test.go

```diff
@@ -1361,6 +1421,334 @@ func TestAllocator(t *testing.T) {
+							}},
+						),
+						subRequest(subReq1, classA, 1, resourceapi.DeviceSelector{
+							CEL: &resourceapi.CELDeviceSelector{
+								Expression: fmt.Sprintf(`device.capacity["%s"].memory.compareTo(quantity("1Gi")) >= 0`, driverA),
+							}},
+						),
+					),
+				),
+			),
+			classes: objects(class(classA, driverA)),
+			slices: objects(
+				slice(slice1, node1, pool1, driverA,
+					device(device1,
+						map[resourceapi.QualifiedName]resource.Quantity{
+							"memory": resource.MustParse("8Gi"),
+						},
+						map[resourceapi.QualifiedName]resourceapi.DeviceAttribute{
+							"driverVersion": {VersionValue: ptr.To("1.0.0")},
+						},
+					),
+				),
+				slice(slice2, node1, pool2, driverA,
+					device(device2,
+						map[resourceapi.QualifiedName]resource.Quantity{
+							"memory": resource.MustParse("2Gi"),
+						},
+						map[resourceapi.QualifiedName]resourceapi.DeviceAttribute{
+							"driverVersion": {VersionValue: ptr.To("2.0.0")},
+						},
+					),
+				),
+			),
+			node: node(node1, region1),
+
+			expectResults: []any{allocationResult(
+				localNodeSelector(node1),
+				deviceAllocationResult(req0, driverA, pool1, device1, false),
+				deviceAllocationResult(req1SubReq1, driverA, pool2, device2, false),
+			)},
+		},
+		"prioritized-list-subrequests-with-allocation-mode-all": {
+			prioritizedList: true,
+			claimsToAllocate: objects(
+				func() *resourceapi.ResourceClaim {
+					claim := claimWithRequests(claim0, nil,
+						requestWithPrioritizedList(req0,
+							subRequest(subReq0, classA, 1),
+							subRequest(subReq1, classA, 1),
+						),
+					)
+					claim.Spec.Devices.Requests[0].FirstAvailableOf[0].AllocationMode = resourceapi.DeviceAllocationModeAll
+					claim.Spec.Devices.Requests[0].FirstAvailableOf[0].Count = 0
+					return claim
+				}(),
+			),
+			classes: objects(class(classA, driverA)),
+			slices: objects(
+				slice(slice1, node1, pool1, driverA,
+					device(device1, nil, nil),
+					device(device2, nil, nil),
+				),
+			),
+			allocatedDevices: []DeviceID{
+				MakeDeviceID(driverA, pool1, device1),
+			},
+			node: node(node1, region1),
+
+			expectResults: []any{allocationResult(
+				localNodeSelector(node1),
+				deviceAllocationResult(req0SubReq1, driverA, pool1, device2, false),
+			)},
+		},
+		"prioritized-list-disabled": {
+			prioritizedList: false,
+			claimsToAllocate: objects(
+				func() *resourceapi.ResourceClaim {
+					claim := claimWithRequests(claim0, nil,
+						requestWithPrioritizedList(req0,
+							subRequest(subReq0, classA, 1),
+						),
+					)
+					return claim
+				}(),
+			),
+			classes: objects(class(classA, driverA)),
+			slices:  objects(sliceWithOneDevice(slice1, node1, pool1, driverA)),
+			node:    node(node1, region1),
+
+			expectResults: nil,
+			expectError:   gomega.MatchError(gomega.ContainSubstring("claim claim-0, request req-0: has subrequests, but the feature is disabled")),
+		},
```

Comment:
We need test cases for backtracking. The resource slice in all of these must have one "large" and one "small" device.

- "no backtracking"
  - request 0 for "large" device"
  - request 1 with two sub-request:
    - sub-request 0 for "small" device
    - sub-request 1 for "large" device
- "no backtracking with sub-request first"
  - request 0 with two sub-request:
    - sub-request 0 for "small" device
    - sub-request 1 for "large" device
  - request 1 for "large" device
- "backtracking for sub-request"
  - request 0 with two sub-request:
    - sub-request 0 for "large" device
    - sub-request 1 for "small" device
  - request 1 for "large" device"

If you want to get fancy, do the same across claims. It just doesn't make a substantial difference.

Reply:
We haver unit tests for all these cases. I will review and incorporate the suggested changes in the allocator, but resolving this comment.